File: lib/data/providers/prayer_data_provider.dart
import '../../domain/entities/mosque.dart';
import '../../domain/entities/prayer_times.dart';
import '../../domain/entities/geo_location.dart';

/// Exception thrown when provider operation fails
class ProviderException implements Exception {
  final String message;
  final int? statusCode;
  final String? providerId;

  const ProviderException({
    required this.message,
    this.statusCode,
    this.providerId,
  });

  @override
  String toString() => 'ProviderException[$providerId]: $message';
}

/// Configuration schema for provider settings
class ConfigField {
  final String key;
  final String label;
  final ConfigFieldType type;
  final bool required;
  final String? description;
  final String? defaultValue;
  final List<String>? options;

  const ConfigField({
    required this.key,
    required this.label,
    required this.type,
    this.required = false,
    this.description,
    this.defaultValue,
    this.options,
  });

  Map<String, dynamic> toJson() => {
    'key': key,
    'label': label,
    'type': type.name,
    'required': required,
    'description': description,
    'defaultValue': defaultValue,
    'options': options,
  };
}

enum ConfigFieldType {
  string,
  password,
  number,
  boolean,
  url,
  select,
}

/// Abstract interface for all prayer data providers
/// Implementations: Provider A (Official API), Provider B (Community Wrapper), Provider C (Scraping)
abstract class PrayerDataProvider {
  /// Provider unique identifier
  String get id;
  
  /// Provider display name
  String get name;
  
  /// Provider description
  String get description;
  
  /// Configuration schema for settings UI
  List<ConfigField> get configSchema;
  
  /// Initialize provider with configuration
  Future<void> initialize(Map<String, dynamic> config);
  
  /// Search mosques by query string
  Future<List<Mosque>> searchMosques(String query, {GeoLocation? location});
  
  /// Get mosques near a location
  Future<List<Mosque>> getNearbyMosques(GeoLocation location, {double radiusKm = 10});
  
  /// Fetch prayer times for a specific mosque
  Future<PrayerTimes> getPrayerTimes(String mosqueId, {DateTime? date});
  
  /// Test connectivity with current configuration
  Future<bool> testConnection();
  
  /// Get mosque details
  Future<Mosque> getMosqueDetails(String mosqueId);
  
  /// Dispose resources
  void dispose();
}

/// Factory for creating providers
class PrayerDataProviderFactory {
  static final Map<String, PrayerDataProvider Function()> _providers = {};

  static void register(String id, PrayerDataProvider Function() creator) {
    _providers[id] = creator;
  }

  static PrayerDataProvider? create(String id) {
    final creator = _providers[id];
    return creator?.call();
  }

  static List<String> get availableProviders => _providers.keys.toList();
}

File: src-tauri/src/providers/prayer_data_provider.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::models::*;

/// Error type for provider operations
#[derive(Debug, thiserror::Error)]
pub enum ProviderError {
    #[error("Network error: {0}")]
    Network(String),
    #[error("Server error: {status_code} - {message}")]
    Server { status_code: u16, message: String },
    #[error("Parse error: {0}")]
    Parse(String),
    #[error("Not found: {0}")]
    NotFound(String),
    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),
    #[error("Other: {0}")]
    Other(String),
}

pub type ProviderResult<T> = Result<T, ProviderError>;

/// Abstract interface for all prayer data providers
#[async_trait]
pub trait PrayerDataProvider: Send + Sync {
    /// Provider unique identifier
    fn id(&self) -> &str;

    /// Provider display name
    fn name(&self) -> &str;

    /// Provider description
    fn description(&self) -> &str;

    /// Configuration schema for settings UI
    fn config_schema(&self) -> Vec<ConfigField>;

    /// Initialize provider with configuration
    async fn initialize(&mut self, config: Value) -> ProviderResult<()>;

    /// Search mosques by query string
    async fn search_mosques(&self, query: &str, location: Option<&GeoLocation>) -> ProviderResult<Vec<Mosque>>;

    /// Get mosques near a location
    async fn get_nearby_mosques(&self, location: &GeoLocation, radius_km: f64) -> ProviderResult<Vec<Mosque>>;

    /// Fetch prayer times for a specific mosque
    async fn get_prayer_times(&self, mosque_id: &str, date: Option<chrono::NaiveDate>) -> ProviderResult<PrayerTimes>;

    /// Test connectivity with current configuration
    async fn test_connection(&self) -> ProviderResult<ProviderTestResult>;

    /// Get mosque details
    async fn get_mosque_details(&self, mosque_id: &str) -> ProviderResult<Mosque>;
}

/// Provider factory
pub struct ProviderFactory;

impl ProviderFactory {
    pub fn create_official_api() -> Box<dyn PrayerDataProvider> {
        Box::new(crate::providers::OfficialApiProvider::new())
    }

    pub fn create_community_wrapper() -> Box<dyn PrayerDataProvider> {
        Box::new(crate::providers::CommunityWrapperProvider::new())
    }

    pub fn create_scraping() -> Box<dyn PrayerDataProvider> {
        Box::new(crate::providers::ScrapingProvider::new())
    }
}
